### PR TITLE
Xavierclements/develop

### PR DESCRIPTION
**Pull request by xavierclements**
- Added functionality to create/check/destroy contexts and commands
- TODO: Parameterize at least the path to the command file

this is an internal review of the changeset prior to opening a PR on `databricks/sbt-databricks` to merge changes upstream

### CODE REVIEW

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 1]
```diff
@@ -135,6 +137,147 @@ private[sbtdatabricks] class DatabricksHttp(endpoint: String, client: HttpClient
...
 
   /**
+   * Create an execution context
+   * @param language the relevant coding language
+   * @param cluster the relevant cluster within which the context will be created
+   * @return The id of the execution context
+   *
+   */
+  private[sbtdatabricks] def createContext(language: String, cluster: Cluster): ContextId = {
+    println(s"Creating '${language}' execution context on cluster '${cluster.name}'")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: use sbt logger

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 2]
```diff
@@ -135,6 +137,147 @@ private[sbtdatabricks] class DatabricksHttp(endpoint: String, client: HttpClient
...
+    val post = new HttpPost(endpoint + CONTEXT_CREATE)
+    val form = new StringEntity(s"""{"language":"${language}","clusterId":"${cluster.id}"}""")
+    form.setContentType("application/json")
+    post.setEntity(form)
+    val response = client.execute(post)
+    val handler = new BasicResponseHandler()
+    val responseString = handler.handleResponse(response).trim
+    println(s"Received the following response: '${responseString}'")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala:156]: logger

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 3]
```diff
@@ -135,6 +137,147 @@ private[sbtdatabricks] class DatabricksHttp(endpoint: String, client: HttpClient
...
+  }
+
+  /**
+   * Check status of an execution context
+   * @param contextId Contains the id of the execution context
+   * @param cluster the relevant cluster
+   * @return status of the execution context
+   */
+  private[sbtdatabricks] def checkContext(contextId: ContextId, cluster: Cluster): ContextStatus = {
+    println(s"Checking execution context '${contextId.id}' on cluster '${cluster.name}'")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala:167]: logger

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 4]
```diff
@@ -135,6 +137,147 @@ private[sbtdatabricks] class DatabricksHttp(endpoint: String, client: HttpClient
...
+  }
+
+  /**
+   * Destroy an execution context
+   * @param contextId Contains the id of the execution context
+   * @param cluster the relevant cluster
+   * @return the id of the execution context
+   */
+  private[sbtdatabricks] def destroyContext(contextId: ContextId, cluster: Cluster): ContextId = {
+    println(s"Terminating execution context '${contextId.id}' on cluster '${cluster.name}'")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala:186]: logger

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 5]
```diff
@@ -135,6 +137,147 @@ private[sbtdatabricks] class DatabricksHttp(endpoint: String, client: HttpClient
...
+    val post = new HttpPost(endpoint + CONTEXT_DESTROY)
+    val form = new StringEntity(s"""{"clusterId":"${cluster.id}","contextId":"${contextId.id}"}""")
+    form.setContentType("application/json")
+    post.setEntity(form)
+    val response = client.execute(post)
+    val handler = new BasicResponseHandler()
+    val responseString = handler.handleResponse(response).trim
+    println(s"Received the following response: '${responseString}'")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala:194]: logger

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 6]
```diff
@@ -135,6 +137,147 @@ private[sbtdatabricks] class DatabricksHttp(endpoint: String, client: HttpClient
...
+   * @param command The code to be executed on the cluster
+   * @return The id of the command
+   *
+   */
+  private[sbtdatabricks] def executeCommand(
+      language: String,
+      cluster: Cluster,
+      contextId: ContextId,
+      commandFilePath: String): CommandId = {
+    println(s"Executing '${language}' command on cluster '${cluster.name}'")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala]: logger

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 7]
```diff
@@ -135,6 +137,147 @@ private[sbtdatabricks] class DatabricksHttp(endpoint: String, client: HttpClient
...
+   * @param contextId The id of the execution context
+   * @param command The code to be executed on the cluster
+   * @return The status of the command
+   *
+   */
+  private[sbtdatabricks] def checkCommand(
+      cluster: Cluster,
+      contextId: ContextId,
+      commandId: CommandId): CommandStatus = {
+    println(s"Checking status of command '${commandId.id}' on cluster '${cluster.name}'")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala:243]: logger

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 8]
```diff
@@ -135,6 +137,147 @@ private[sbtdatabricks] class DatabricksHttp(endpoint: String, client: HttpClient
...
+    val form =
+      URLEncodedUtils.format(List(new BasicNameValuePair("clusterId", cluster.id),
+                                  new BasicNameValuePair("contextId", contextId.id),
+                                  new BasicNameValuePair("commandId", commandId.id)), "utf-8")
+    val request = new HttpGet(endpoint + COMMAND_STATUS + "?" + form)
+    val response = client.execute(request)
+    val handler = new BasicResponseHandler()
+    val responseString = handler.handleResponse(response).trim
+    println(s"Received the following response: '${responseString}'")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala:252]: logger

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 9]
```diff
@@ -135,6 +137,147 @@ private[sbtdatabricks] class DatabricksHttp(endpoint: String, client: HttpClient
...
+   * @param contextId The id of the execution context
+   * @param command The code to be executed on the cluster
+   * @return The id of the command
+   *
+   */
+  private[sbtdatabricks] def cancelCommand(
+      cluster: Cluster,
+      contextId: ContextId,
+      commandId: CommandId): CommandId = {
+    println(s"Cancelling command '${commandId.id}' on cluster '${cluster.name}'")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala:269]: logger

File: src/main/scala/sbtdatabricks/DatabricksHttp.scala [hunk 10]
```diff
@@ -135,6 +137,147 @@ private[sbtdatabricks] class DatabricksHttp(endpoint: String, client: HttpClient
...
+    val post = new HttpPost(endpoint + COMMAND_CANCEL)
+    val form = new StringEntity(s"""{"clusterId":"${cluster.id}","contextId":"${contextId.id}","commandId":"${commandId.id}"}""")
+    form.setContentType("application/json")
+    post.setEntity(form)
+    val response = client.execute(post)
+    val handler = new BasicResponseHandler()
+    val responseString = handler.handleResponse(response).trim
+    println(s"Received the following response: '${responseString}'")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksHttp.scala:277]: logger

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 11]
```diff
@@ -24,6 +24,9 @@ object DatabricksPlugin extends AutoPlugin {
     val dbcLibraryPath = settingKey[String]("Where in the workspace to add the libraries.")
     val dbcListClusters = taskKey[Unit]("List all available clusters and their states.")
     val dbcRestartClusters = taskKey[Unit]("Restart the given clusters.")
+    val dbcExecuteCommand = taskKey[Unit]("Execute a command on a particular cluster")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: This should return something else than `Unit` eg `ExecutionResult`

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 12]
```diff
@@ -24,6 +24,9 @@ object DatabricksPlugin extends AutoPlugin {
...
+    val dbcCommandFile = settingKey[String]("Location of file containing the command to be executed")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala:31]: `java.nio.Path` ?

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 13]
```diff
@@ -24,6 +24,9 @@ object DatabricksPlugin extends AutoPlugin {
...
+    val dbcExecutionLanguage = settingKey[String]("Which language is to be used when executing a command")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]:
``` scala
sealed trait SparkProgrammingLanguage { def is: String }
case object Scala extends SparkProgrammingLanguage { def is = "scala" }
case object Python extends SparkProgrammingLanguage { def is = "python" }
case object Java extends SparkProgrammingLanguage { def is = "java" }
```

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 14]
```diff
@@ -124,7 +127,57 @@ object DatabricksPlugin extends AutoPlugin {
       }
     }
   }
-  
+
+  private lazy val executeCommandImpl: Def.Initialize[Task[Unit]] = Def.taskDyn {
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: `Unit` => `ExecutionResult`

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 15]
```diff
@@ -124,7 +127,57 @@ object DatabricksPlugin extends AutoPlugin {
...
+    val allClusters = dbcFetchClusters.value
+    val commandFile = dbcCommandFile.value
+    Def.task {
+      for (clusterName <- onClusters) {
+        // Duplicating aspects of code in DatabricksHttp.foreachCluster
+        // - this is due to inability to pass values between DatabricksHttp function
+        // calls within the foreachCluster
+        val givenCluster = allClusters.find(_.name == clusterName)
+        if (givenCluster.isEmpty) {
+          throw new NoSuchElementException(s"Cluster with name: '${clusterName} not found!")
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala:148]: Not sure:
`logger.error` + http://www.scala-sbt.org/0.12.2/docs/Detailed-Topics/Tasks.html#handling-failure

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 16]
```diff
@@ -183,3 +237,9 @@ case class LibraryStatus(
     attachAllClusters: Boolean,
     statuses: List[LibraryClusterStatus])
 case class LibraryClusterStatus(clusterId: String, status: String)
+case class ContextId(id: String)
+case class ContextStatus(status: String, id: String)
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala:247]: status should be an ADT
see https://github.com/ypg-data/sbt-databricks/pull/1/files#r30408327

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 17]
```diff
@@ -124,7 +127,57 @@ object DatabricksPlugin extends AutoPlugin {
...
+          throw new NoSuchElementException(s"Cluster with name: '${clusterName} not found!")
+        }
+        val confirmedCluster = givenCluster.get
+        // parameterize the code / command that is issued here
+        // Attempt to create an execution context on the relevant cluster
+        var contextCorrect = false
+        val contextId = client.createContext(language, confirmedCluster)
+        while (!contextCorrect) {
+          val contextStatus = client.checkContext(contextId, confirmedCluster)
+          if (contextStatus.status == "Running") {
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: see https://github.com/ypg-data/sbt-databricks/pull/1/files#r30408647

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 18]
```diff
@@ -124,7 +127,57 @@ object DatabricksPlugin extends AutoPlugin {
...
+            contextCorrect = true
+          } else if (contextStatus.status == "Error") {
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: see https://github.com/ypg-data/sbt-databricks/pull/1/files#r30408647

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 19]
```diff
@@ -124,7 +127,57 @@ object DatabricksPlugin extends AutoPlugin {
...
+            throw new Exception(s"The Execution Context '${contextId.id} returned an error. Terminated context")
+          }
+        }
+
+        // Attempt to issue a command
+        val commandId = client.executeCommand(language, confirmedCluster, contextId, commandFile)
+        var commandComplete = false
+        while (!commandComplete) {
+          val commandStatus = client.checkCommand(confirmedCluster, contextId, commandId)
+          if (commandStatus.status == "Finished") {
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: see https://github.com/ypg-data/sbt-databricks/pull/1/files#r30408647

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala [hunk 20]
```diff
@@ -124,7 +127,57 @@ object DatabricksPlugin extends AutoPlugin {
...
+            commandComplete = true
+          } else if (commandStatus.status == "Error") {
```
Comment (MasseGuillaume) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala]: see https://github.com/ypg-data/sbt-databricks/pull/1/files#r30408647

**Comment by suhailshergill** [comment]
@MasseGuillaume makes valid points about logging etc. but let's keep our changes/improvements to a minimum for this specific PR. the main goal of this PR is for internal approval. after which a PR needs to be made to ypg-databricks. i don't want us to spend time making changes if that's going to cause issues with getting this changeset merged into upstream

**Comment by suhailshergill** [comment]
@xavierclements nothing too offensive, has my +1. @MasseGuillaume 's recommendation of having something other than strings makes sense, and if it's lightweight enough to implement ok with us doing it prior to merge here. otherwise that and other related issues seem better to be brought up in PR on `databricks/sbt-databricks` repo

**Comment by xavierclements** [comment]
@suhailshergill @MasseGuillaume ok I'll keep it to removing the strings. though the logging would help us as it would be present in our builds.

File: src/main/scala/sbtdatabricks/DatabricksPlugin.scala (hunk 16 above)
Comment (xavierclements) [inline src/main/scala/sbtdatabricks/DatabricksPlugin.scala:247]: @MasseGuillaume this is the return type of the Commands. I've ripped out the comparisons to these responses to an ADT.

**Comment by xavierclements** [comment]
@suhailshergill @MasseGuillaume removed all strings save for the File type. It was causing problems. I'll  put a todo (https://github.com/ypg-data/sbt-databricks/commit/97e89e41018e60d6e63551a432deb54b72ba7f7d)

**Pull request merged**